Some other changes

cage2 opened the pull request:
Same more changes: bugfix and a little new feature! :)
Bye!
C.

Commit:
- added a  bit of (very simple)  heuristic to place annotation  in the
  right  position  even  if  the  file has  been  saved  when  not  in
  annotation mode (i.e. when the fingerprints mismatch).
  Do not expect anything smart, though;
- remove annotation when the annotated text is deleted from the buffer;
- stilistic changes:
  uses '(null ...' instead of '(eq nil ...'
  uses "annotationp" instead of "(get overlay 'annotation)";
- use  library  functions instead  of  direct  access (via  'nth')  of
  annotation database.

bastibe commented on annotate.el:
```diff
@@ -871,6 +879,14 @@ essentially what you get from:
 (annotate-annotations-from-dump (annotate-load-annotations))). "
   (nth 2 annotation))
 
+(defun annotate-sample-text-of-annotation (annotation)
```
These function names are confusing. I think we should find a better naming scheme for `text-of-annotation` and `sample-text-of-annotation`. 

Maybe `annotate-destination-text-of-annotation` and `annotate-annotation-text-of-annotation`?

It's not perfect yet...

cage2 replied:
So if i understand correctly:

``` text
 from 'annotate-sample-text-of-annotation' to  'annotate-annotation-text-of-annotation'
 from 'annotate-text-of-annotation'        to  'annotate-destination-text-of-annotation'
```
right?

bastibe replied:
Yeah, but now that you write them out, those are not good names yet.

So we have to differentiate between the text being annotated, and the annotation text. Maybe that's the key right there: "annotation text" vs. "annotated text". And maybe mix up that "text" to make them more distinct:

    from annotate-sample-text-of-annotation   to annotate-annotated-text
    from annotate-text-of-annotation          to annotate-annotation-string

(And possibly reword other occurrences accordingly)

What do you think about that?

cage2 replied:
Hi bastibe, and welcome back! :-)

I agree with the changes and i am going to push changes accordingly to your suggestions. I have to admit i am a bit confused between annotated text and annotation text i hope i did right with this changes. :-D

Bye!
 C.

bastibe commented on annotate.el:
```diff
@@ -950,18 +968,65 @@ essentially what you get from:
+                                 (let ((highlight (make-overlay start end)))
+                                   (overlay-put highlight 'face 'annotate-highlight)
+                                   (overlay-put highlight 'annotation annotation)))
+              (move-lines        (start line-count)
```
How about `beginning-of-nth-line`?

cage2 replied:
sure! Changed!

bastibe commented on annotate.el:
```diff
@@ -950,18 +968,65 @@ essentially what you get from:
+                                   (forward-line line-count)
+                                   (beginning-of-line)
+                                   (point)))
+              (go-backward       (start)
```
`beginning-of-previous-line`? But why two lines?

bastibe replied:
Oh, I see. It's two lines because that's the "search region" where a moved annotation can be, right? It might be useful to make that a customizable variable.

cage2 replied:
Good idea!

cage2 replied:
> (I'd have appreciated a more descriptive pull request text ;-) )

Sorry!

bastibe commented on annotate.el:
```diff
@@ -950,18 +968,65 @@ essentially what you get from:
+                                   (point)))
+              (go-backward       (start)
+                                 (move-lines start -2))
+              (go-forward        (start)
```
`beginning-of-next-line`? But why two lines?

bastibe reviewed (changes requested):
Cool! Finally, we'll have some kind of plausibility check that the annotations are still valid.

(I'd have appreciated a more descriptive pull request text ;-) )

bastibe commented on annotate.el:
```diff
@@ -950,18 +968,65 @@ essentially what you get from:
+      (string= "" a)))
+
+(defun annotate-create-annotation (start end &optional text sample)
   "Create a new annotation for selected region."
```
Docstring probably needs to be expanded with new functionality.

cage2 replied:
I tried, not very happy with the results, though! :(

bastibe replied:
Iif I'm reading the code correctly, it does

- If no text or sample are given, create a new annotation between start and end, and ask for the annotation text
- If a text is given but no sample, create a new annotation between start and end
- If both text and sample are given, and the sample matches the text between start and end, create a new annotation between start and end
- If both text and sample are given, and the sample does not match the text between start and end, search the surrounding area for a matching location, and create a new annotation there
- If both text and sample are given, and the sample does not match the text between start and end, and no matching location is nearby, silently ignore the annotation

It might be useful to split this into multiple functions (as a general rule of thumb, I try to split stuff into multiple functions if I start to feel compelled to add an "and" to the function name).

For example, we could make `text` and `sample` non-optional by changing `annotate-annotate` to grab them before calling `annotate-create-annotation`.

And then maybe add a failure condition for non-matching annotations, probably as a warning.

cage2 replied:
Hi!
Yes i think you understand correctly the goal of the function (and i hope my code works this way :-)) 

Unfortunately i think i do not understand very well how to refactor the function, instead. :)
The two optional arguments in ``annotate-create-annotation`` are used only when the annotations are restored from the database i can not see how they can be used in `annotate-annotate`, should we change: `(annotate-create-annotation start end)` to `(annotate-create-annotation start end nil nil)`? Please help me to understand.

Anyway, good idea for the warning if annotation matching fails! I like it! :)

Bye!

bastibe replied:
Hi!

When looking through the code, I noticed that all but one place calls `annotate-create-annotation` with four arguments. If that one place, `annotate-annotate`, would use all four arguments, we could make the last two arguments non-optional. So I propose to change `annotate-annotate` to supply the last two arguments, and to make them non-optional. This will clean up the implementation of `annotate-create-annotation` somewhat, because we won't have to include all the checks for the last two arguments any longer.

Does that explain it better?

cage2 replied:
Hello!!

yes i think the fog in my brain is starting to going away! :-)

so, i just made the two arguments non optional and now i should move all that stuff:

``` lisp 
   (let ((annotation (or annotation-text
                          (read-from-minibuffer "Annotation: "))))
      (when (not (or (null annotation) 
                     (string= "" annotation))) ; note this can be changed with
                                               ; (not (string-empty-p annotation))
        (if (not (annotate-string-empty-p annotated-text))
 ```
up to `annotate-annotate`, is it?

Bye!
C.

bastibe replied:
That's the idea!

Except of course you won't need all the conditionals, as `annotate-annotate` knows it has to ask for the annotation-string and can grab the annotated-text from the buffer.

Commit:
- changed function name: 'move-lines' -> 'beginning-of-nth-line'.

Commit:
- using a customizable variable for region where to look for annotated
  text.

Commit:
- try to elaborate a bit more the function goal.

Commit:
- try to elaborate more a function goal.

Commit:
- fixed region's limit to break an annotation;
- binding 'print-length' to nil before dumping annotation because you
  never can tell; :)
- fixed test to ellipsize text in summary window.

Commit:
- changed two function names and reworded 'annotate-create-annotation' docstring

  in details:

  from 'annotate-sample-text-of-annotation'   to 'annotate-annotated-text'
  from 'annotate-text-of-annotation'          to 'annotate-annotation-string'

Commit:
- made two arguments of annotate-create-annotation non optional;
- cosmetic changes to a single docstring.

Commit:
- moving  some   args  checks  from   'annotate-create-annotation'  to
  'annotate-annotate' where they do make more sense;
- fixed variable scoping in a local funcion;
- expanded a bit a docstring.

cage2 commented:
On Thu, Nov 07, 2019 at 11:19:21AM -0800, Bastian Bechtold wrote:

Hello!


>
> That's the idea!
> Except  of  course   you  won't  need  all   the  conditionals,  as
> `annotate-annotate` knows  it has  to ask for  the annotation-string
> and can grab the annotated-text from the buffer.

Very well!

I think (an hope) this commit:

https://github.com/bastibe/annotate.el/pull/47/commits/3108c44c5308f6d26e3cb143bf38cc69e300298a

should implement your suggestion.

To me this change does make a lot of sense!

Bye!
C.

bastibe commented:
Absolutely! Great work!

Commit:
- fixed reference  to a  non-existent variable  in the  procedure that
  load the metadata (old format only).
- given more  meaningful name to  two variables in the  procedure that
  load the metadata.

bastibe commented:
As far as I can tell, this is almost complete! I'd just add an error message if the annotation couldn't be loaded and then this looks good to merge!

cage2 commented:
On Fri, Nov 08, 2019 at 06:46:32AM -0800, Bastian Bechtold wrote:
> Absolutely! Great work!

Thank you very much! :)

Bye
C.

cage2 commented:
On Fri, Nov 08, 2019 at 06:49:59AM -0800, Bastian Bechtold wrote:

Hi!

> As far as I can tell, this is almost complete! I'd just add an error
> message if the annotation couldn't be loaded

Sure, i am going to add it, seems to me it should not be a complicated
goal to accomplish so i hope to add int in the next hour.

> and then this looks good to merge!

Also  i would  like  to prevent  to  show a  summary  window when  the
metadata db is  empty or contain no actual annotations,  when the former
happens (empty window) this looks a bit not elegant to me.  In this case
i would prefer just to print a message to inform the user that there are no
annotations to show.  Are you OK with adding this feature?

Bye!!

C.

Commit:
- added a warning when searching for a annotated text fails.

Commit:
- fixed docstring.

Commit:
- removed useless quotation mark.

Commit:
- prevent showing of summary window if metadata contains no annotations at all.

cage2 commented:
Well, seems i have pushed the changes for the checks about the summery window, sorry! 
Anyway if you do not like it just tell me and i will revert the last commit.

Bye!!!
C.

Commit:
- refactored 'annotate-actual-file-name' and used where appropriate.

bastibe commented:
Cool, I like it!

Thank you very much for another great contribution.

Feel free to merge when you're ready!

Commit:
- fixed macro: "annotate-with-inhibit-modification-hooks".

cage2 commented:
On Sat, Nov 09, 2019 at 04:44:59AM -0800, Bastian Bechtold wrote:
> Cool, I like it!
>
> Thank you very much for another great contribution.
>
> Feel free to merge when you're ready!

With fixing this macro i think we can merge the two branches.

Bye!
C.

cage2 commented:
An that's all, for now! :D

bastibe commented:
🎉
